src/storage/src/hummock/utils.rs

```diff
@@ -166,26 +167,47 @@ pub fn prune_nonoverlapping_ssts<'a>(
     ssts[start_table_idx..=end_table_idx].iter()
 }
 
-#[derive(Debug)]
+type RequestQueue = VecDeque<(Sender<MemoryTrackerImpl>, u64)>;
+
 struct MemoryLimiterInner {
     total_size: AtomicU64,
-    notify: Notify,
+    controller: Mutex<RequestQueue>,
+    pending_request_count: Arc<AtomicU64>,
     quota: u64,
+    fast_quota: u64,
 }
 
 impl MemoryLimiterInner {
     fn release_quota(&self, quota: u64) {
         self.total_size.fetch_sub(quota, AtomicOrdering::Release);
-        self.notify.notify_waiters();
     }
 
     fn add_memory(&self, quota: u64) {
         self.total_size.fetch_add(quota, AtomicOrdering::SeqCst);
     }
 
+    fn may_notify_waiters(self: &Arc<Self>) {
+        let mut waiters = self.controller.lock();
+        while let Some((tx, quota)) = waiters.pop_front() {
+            if !self.try_require_memory_in_capacity(quota, self.quota) {
+                waiters.push_front((tx, quota));
+                break;
+            }
+            let _ = tx.send(MemoryTrackerImpl::new(self.clone(), quota));
```

Comment:
IIUC, the difference between `MemoryTracker` and `MemoryTrackerImpl` is that, `MemoryTrackerImpl` only releases the quota, but `MemoryTracker` will also notify waiters. 

If so, I've come up with a corner case that may cause endless await.

Let's say the capacity is 100, and we have two threads, A wants 100 and B wants 100, and both of them are currently in the queue. 

1. When someone else releases the quota and call `may_notify_waiters`, it will find A in the queue, and create a `MemoryTrackerImpl` with quota 100 and send to A via the tx. The `send` method returns successfully and the `MemoryTrackerImpl` is added to the buffer. 
2. However, before A poll the rx and take the `MemoryTrackerImpl` out, the future of A may be cancelled and dropped, and then only the `MemoryTrackerImpl` gets dropped, which only adds the quota back to the atomic variable, but won't notify waiters in the queue (which is B). 
3. By now, B is waiting for notification, but no one is going to release any quota and notify B, which causes endless await.

Reply:
good catch.
I will design a guard to notify other thread when this future is canceled.